pkg/execution/state/redis_state/cluster.go

```diff
@@ -0,0 +1,107 @@
+package redis_state
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"github.com/redis/rueidis"
+	"runtime"
+	"time"
+)
+
+type RetriableClient interface {
+	Do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed) (resp rueidis.RedisResult)
+}
+
+type noopRetriableClient struct {
+	r rueidis.Client
+}
+
+func (r noopRetriableClient) B() rueidis.Builder {
+	return r.r.B()
+}
+
+func (r noopRetriableClient) Do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed) (resp rueidis.RedisResult) {
+	return r.r.Do(ctx, cmd(r.r))
+}
+
+func newNoopRetriableClient(client rueidis.Client) RetriableClient {
+	return noopRetriableClient{client}
+}
+
+type retryClusterDownClient struct {
+	r rueidis.Client
+}
+
+func (r retryClusterDownClient) B() rueidis.Builder {
+	return r.r.B()
+}
+
+func (r retryClusterDownClient) do(ctx context.Context, cmd func(client rueidis.Client) rueidis.Completed, attempts int) rueidis.RedisResult {
+	resp := r.r.Do(ctx, cmd(r.r))
```

Comment:
Unfortunately, Rueidis cannot reuse commands unless Pinned (in which case the commands are not recycled properly). For this reason, we needed to change the signature of cmd to expect a function which is executed for each attempt.